container/multi-value-slice/multi_value_slice.go

```diff
@@ -0,0 +1,380 @@
+type Slice[V comparable, O interfaces.Identifiable] struct {
+	sharedItems     []V
+	individualItems map[uint64]*MultiValue[V]
+	appendedItems   []*MultiValue[V]
+	cachedLengths   map[interfaces.Id]int
+	lock            sync.RWMutex
+}
+
+// Init initializes the slice with sensible defaults. Input values are assigned to shared items.
+func (s *Slice[V, O]) Init(items []V) {
+	s.sharedItems = items
+	s.individualItems = map[interfaces.Id]*MultiValue[V]{}
+	s.appendedItems = []*MultiValue[V]{}
+	s.cachedLengths = map[interfaces.Id]int{}
+}
+
+// Len returns the number of items for the input object.
+func (s *Slice[V, O]) Len(obj O) int {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	l, ok := s.cachedLengths[obj.Id()]
+	if !ok {
+		return len(s.sharedItems)
+	}
+	return l
+}
+
+// Copy copies items between the source and destination.
+func (s *Slice[V, O]) Copy(src O, dst O) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, item := range s.individualItems {
+		for _, v := range item.Values {
+			found := compareIds(v.ids, src.Id(), func(_ int) {
+				v.ids = append(v.ids, dst.Id())
+			})
+			if found {
+				break
+			}
+		}
+	}
+
+appendedLoop:
+	for _, item := range s.appendedItems {
+		found := false
+		for _, v := range item.Values {
+			found = compareIds(v.ids, src.Id(), func(_ int) {
+				v.ids = append(v.ids, dst.Id())
+			})
+			if found {
+				break
+			}
+		}
+		if !found {
+			// This is an optimization. If we didn't find an appended item at index i,
+			// then all larger indices don't have an appended item for the object either.
+			break appendedLoop
+		}
+	}
+
+	srcLen, ok := s.cachedLengths[src.Id()]
+	if ok {
+		s.cachedLengths[dst.Id()] = srcLen
+	}
+}
+
+// Value returns all items for the input object.
+func (s *Slice[V, O]) Value(obj O) []V {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	result := make([]V, len(s.sharedItems))
+	for i, item := range s.sharedItems {
+		ind, ok := s.individualItems[uint64(i)]
+		if !ok {
+			result[i] = item
+		} else {
+			found := false
+			for _, v := range ind.Values {
+				found = compareIds(v.ids, obj.Id(), func(_ int) {
+					result[i] = v.val
+				})
+				if found {
+					break
+				}
+			}
+			if !found {
+				result[i] = item
+			}
+		}
+	}
+
+	for _, item := range s.appendedItems {
+		found := false
+		for _, v := range item.Values {
+			found = compareIds(v.ids, obj.Id(), func(_ int) {
+				result = append(result, v.val)
+			})
+			if found {
+				break
+			}
+		}
+		if !found {
+			// This is an optimization. If we didn't find an appended item at index i,
+			// then all larger indices don't have an appended item for the object either.
+			return result
+		}
+	}
+
+	return result
+}
+
+// At returns the item at the requested index for the input object.
+// If the object has an individual value at that index, it will be returned. Otherwise the shared value will be returned.
+// If the object has an appended value at that index, it will be returned.
```

Comment:
Something im unclear about is the ordering of return types: why do we return individual first, then shared, then appended in that order?

Reply:
Appended items' indices are always larger than shared/individual items' indices. We first check if the index is within the length of shared items. If it is, then we return an individual value at that index - if it exists - or a shared value otherwise. If the index is beyond the length of shared values, it is an appended item and that's what gets returned.

I will change the description to the above, which is more detailed. And more clear, I hope.